scripts/backtest_sites.py

```diff
@@ -0,0 +1,447 @@
+    t0_datapipe = IterableWrapper(
+        [[t0 for site_id in ALL_SITE_IDS] for t0 in available_target_times]
+    )
+    t0_datapipe = t0_datapipe.sharding_filter()
+    t0_datapipe = t0_datapipe.unbatch(
+        unbatch_level=1
+    )  # might not need this part since the site datapipe is creating examples
+
+    t0_datapipe = t0_datapipe.set_length(num_t0s * len(ALL_SITE_IDS))
+    location_pipe = location_pipe.set_length(num_t0s * len(ALL_SITE_IDS))
+
+    return location_pipe, t0_datapipe
+
+
+class ModelPipe:
+    """A class to conveniently make and process predictions from batches"""
+
+    def __init__(self, model, ds_site: xr.Dataset):
+        """A class to conveniently make and process predictions from batches
+
+        Args:
+            model: PVNet site level model
+            ds_site:xarray dataset of pv site true values and capacities
+        """
+        self.model = model
+        self.ds_site = ds_site
+
+    def predict_batch(self, batch: NumpyBatch) -> xr.Dataset:
+        """Run the batch through the model and compile the predictions into an xarray DataArray
+
+        Args:
+            batch: A batch of samples with inputs for each site for the same init-time
+
+        Returns:
+            xarray.Dataset of all site and national forecasts for the batch
+        """
+        # Unpack some variables from the batch
+        id0 = batch[BatchKey.pv_t0_idx]
+
+        t0 = batch[BatchKey.pv_time_utc].cpu().numpy().astype("datetime64[s]")[0, id0]
+        n_valid_times = len(batch[BatchKey.pv_time_utc][0, id0 + 1 :])
+        model = self.model
+
+        # Get valid times for this forecast
+        valid_times = pd.to_datetime(
+            [t0 + np.timedelta64((i + 1) * FREQ_MINS, "m") for i in range(n_valid_times)]
+        )
+
+        # Get effective capacities for this forecast
+        # site_capacities = ds_site.nominal_capacity_wp.values
+        # Get the solar elevations. We need to un-normalise these from the values in the batch
+        elevation = batch[BatchKey.pv_solar_elevation] * ELEVATION_STD + ELEVATION_MEAN
+        # We only need elevation mask for forecasted values, not history
+        elevation = elevation[:, id0 + 1 :]
+
+        # Make mask dataset for sundown
+        da_sundown_mask = xr.DataArray(
+            data=elevation < MIN_DAY_ELEVATION,
+            dims=["pv_system_id", "target_datetime_utc"],
+            coords=dict(
+                pv_system_id=ALL_SITE_IDS,
+                target_datetime_utc=valid_times,
+            ),
+        )
+
+        with torch.no_grad():
+            # Run batch through model to get 0-1 predictions for all sites
+            device_batch = copy_batch_to_device(batch_to_tensor(batch), device)
+            y_normed_site = model(device_batch).detach().cpu().numpy()
+        da_normed_site = preds_to_dataarray(y_normed_site, model, valid_times, ALL_SITE_IDS)
+
+        # TODO fix this step: Multiply normalised forecasts by capacities and clip negatives
```

Comment:
Whats wrong with it?

Reply:
Seemed to be mixing up which capacities were for which site, haven't dug into this fully yet but might just be a site datapipes issues, left the outputs as a normalised output for now which can be converted to MW in post processing

Reply:
thanks